### DISCUSSION TITLE
Track location in background

### DISCUSSION THREAD

**Discussion by alefcarlos** [IEvangelist/blazorators#52] [Q&A]
Using HostedService to track user location is not a possibility, is there any way to track the location in background using blazor ?

I would like to invoke a SignalR method on `OnPositionRecieved`.

**Comment by IEvangelist** [comment]
Is the idea to have this be something that tracks the position with the server or wasm approach?

**Reply by alefcarlos** [reply]
I was trying to find out the best approach. I think the server was a better solution, but I dont know yet.

Using wasm with `InteractiveWebAssembly`:

I was trying to inject `IGeolocationService` in `App.razor`, but the component is always diposed and it creates a new  `HubConnection` on every initialized. I dont think it is a good solution.


Using server:
I was trying to use IGeolocationService in a HostedService, but it does not work

**Reply by IEvangelist** [reply]
Ok. Yeah, you'd have to use the server package, as the WebAssembly packages do not work with the `InteractiveWebAssembly` newer rendering modes. I don't think you want to manually have your own `HubConnection` either. When you're doing the server rendering modes, there's already a SignalR connection that manages DOM updates on your behalf. I think, you'd simply want to inject the `IGeolocationService` into the `App.razor`, and register the callback and call `WatchPosition`. This will be called everytime the client signals an update. Then, in the callback itself, you're already operating on the server. Then you could do whatever you want with those values. Does that make sense?

**Reply by alefcarlos** [reply]
>  Then, in the callback itself, you're already operating on the server. Then you could do whatever you want with those values. Does that make sense?

I agree. It does. The `App.razor` must implment `IDisposable` to clear the watchId as well, right ?

**Reply by alefcarlos** [reply]
But this will only get the position on the first render, an the callbaback will never be called againa, because of the disposable

**Reply by alefcarlos** [reply]
@IEvangelist, actually, event after the clear the callback is fired. I'm trying this now. Is this the expected behaviour ?

```csharp
protected override async Task OnAfterRenderAsync(bool firstRender)
{
    if (firstRender is false)
    {
        return;
    }

    _watchId =  await Geolocation.WatchPositionAsync(
        component: this,
        onSuccessCallbackMethodName: nameof(OnPositionReceived),
        onErrorCallbackMethodName: nameof(OnPositionError),
        options: _options);
}

[JSInvokable]
public void OnPositionReceived(GeolocationPosition position)
{
    _isLoading = false;
    _position = position;
    StateHasChanged();
}

[JSInvokable]
public void OnPositionError(GeolocationPositionError positionError)
{
    _isLoading = false;
    _positionError = positionError;
    StateHasChanged();
}

public async ValueTask DisposeAsync()
{
    await Geolocation.ClearWatchAsync(_watchId);
}
```

**Reply by alefcarlos** [reply]
I got the point, the `DisposeAsync` was throwing exception:

```
fail: Microsoft.AspNetCore.Server.Kestrel[13]
      Connection id "0HN3EK0DEPNV3", Request id "0HN3EK0DEPNV3:00000005": An unhandled exception was thrown by the application.
      System.InvalidOperationException: JavaScript interop calls cannot be issued at this time. This is because the component is being statically rendered. When prerendering is enabled, JavaScript interop calls can only be performed during the OnAfterRenderAsync lifecycle method.
         at Microsoft.AspNetCore.Components.Server.Circuits.RemoteJSRuntime.BeginInvokeJS(Int64 asyncHandle, String identifier, String argsJson, JSCallResultType resultType, Int64 targetInstanceId)
         at Microsoft.JSInterop.JSRuntime.InvokeAsync[TValue](Int64 targetInstanceId, String identifier, CancellationToken cancellationToken, Object[] args)
         at Microsoft.JSInterop.JSRuntime.InvokeAsync[TValue](Int64 targetInstanceId, String identifier, Object[] args)
         at Microsoft.JSInterop.JSRuntimeExtensions.InvokeVoidAsync(IJSRuntime jsRuntime, String identifier, Object[] args)
         at Hbsa.Acqua.Geo.Components.Pages.Geo.DisposeAsync() in C:\repos\dev.azure.com\hidrovias-brasil\acqa-poc-blazor-embedded\src\gps_server\Hbsa.Acqua.Geo\Components\Pages\Geo.razor.cs:line 53
         at Microsoft.AspNetCore.Components.RenderTree.Renderer.<>c__DisplayClass85_0.<<Dispose>g__HandleAsyncExceptions|1>d.MoveNext()
      --- End of stack trace from previous location ---
         at Microsoft.AspNetCore.Components.RenderTree.Renderer.<>c__DisplayClass85_0.<<Dispose>g__HandleAsyncExceptions|1>d.MoveNext()
      --- End of stack trace from previous location ---
         at Microsoft.AspNetCore.Components.RenderTree.Renderer.DisposeAsync()
         at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.<DisposeAsync>g__Await|26_0(Int32 i, ValueTask vt, List`1 toDispose)
         at Microsoft.AspNetCore.Http.Features.RequestServicesFeature.<DisposeAsync>g__Awaited|9_0(RequestServicesFeature servicesFeature, ValueTask vt)
         at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.<FireOnCompleted>g__ProcessEvents|242_0(HttpProtocol protocol, Stack`1 events)
```

**Reply by IEvangelist** [reply]
Let me try it locally, and report back... I'd expect that to work. If it doesn't, and the App is disposed, you could create a slightly different approach. You could have a background service that listens for changes, and an API that accepts the changes. You could then use session storage with a new `Guid` to act as a client ID that corresponds to location updates.

**Reply by IEvangelist** [reply]
So, locally... I'm testing this now, and it is working as I'd expect. See: https://github.com/IEvangelist/blazorators/blob/main/samples/BlazorServer.ExampleConsumer/Pages/Track.razor.cs. This sample code, is server-based and while the app is running and, on this page, it works as expected. If you're trying to track at the App level, that's not something I've done but I'd be curious why would you need that?

In other words, when your Geo page goes out of scope it is disposed thus stopping the geolocation callbacks.

**Reply by alefcarlos** [reply]
I was using that example. I have at least 5 pages that needs the geolocalization, and I was thinking about creating a backgroundservice to set a singleton instance with the current position and use that singleton into the pages, just to reduce code.

So the best approach is to implement the GetCurrentPosition on each page, right ?
